src/geometry/manifold/instances/sphere.lean

```diff
@@ -0,0 +1,269 @@
+  open_target := is_open_univ,
+  continuous_to_fun := continuous_on_stereo_to_fun.comp continuous_subtype_coe.continuous_on
+    (λ w h, h ∘ subtype.ext ∘ eq.symm ∘ (inner_eq_norm_mul_iff_of_norm_one hv (by simp)).mp),
+  continuous_inv_fun := (continuous_stereo_inv_fun hv).continuous_on }
+
+@[simp] lemma stereographic_source (hv : ∥v∥ = 1) :
+  (stereographic hv).source = {⟨v, by simp [hv]⟩}ᶜ :=
+rfl
+
+@[simp] lemma stereographic_target (hv : ∥v∥ = 1) : (stereographic hv).target = set.univ := rfl
+
+end stereographic_projection
+
+section charted_space
+
+variables [finite_dimensional ℝ E]
+
+/-! ### Charted space structure on the sphere
+
+In this section we construct a charted space structure on the unit sphere in a finite-dimensional
+real inner product space `E`; that is, we show that it is locally homeomorphic to the Euclidean
+space of dimension one less than `E`.
+
+The restriction to finite dimension is for convenience.  The most natural `charted_space`
+structure for the sphere uses the stereographic projection from the antipodes of a point as the
+canonical chart at this point.  However, the codomain of the stereographic projection constructed
+in the previous section is `(ℝ ∙ v)ᗮ`, the orthogonal complement of the vector `v` in `E` which is
+the "north pole" of the projection, so a priori these charts all have different codomains.
+
+So it is necessary to prove that these codomains are all continuously linearly equivalent to a
+fixed normed space.  This could be proved in general by a simple case of Gram-Schmidt
+orthogonalization, but in the finite-dimensional case it follows more easily by dimension-counting.
+-/
+
+/-- Variant of the stereographic projection (see `stereographic`), for the sphere in a finite-
+dimensional inner product space `E`.  This version has codomain the Euclidean space of dimension
+`findim ℝ E - 1`. -/
+def stereographic' (v : sphere (0:E) 1) :
+  local_homeomorph (sphere (0:E) 1) (euclidean_space ℝ (fin (findim ℝ E - 1))) :=
+(stereographic (norm_eq_of_mem_sphere v)).trans
+(continuous_linear_equiv.of_findim_eq
+( begin
+    rw findim_orthogonal_span_singleton (nonzero_of_mem_unit_sphere v),
+    simp
+  end )).to_homeomorph.to_local_homeomorph
+
+@[simp] lemma stereographic'_source (v : sphere (0:E) 1) :
+  (stereographic' v).source = {v}ᶜ :=
+by simp [stereographic']
+
+@[simp] lemma stereographic'_target (v : sphere (0:E) 1) :
+  (stereographic' v).target = set.univ :=
+by simp [stereographic']
+
+/-- The unit sphere in a finite-dimensional inner product space `E` is a charted space modelled on
+the Euclidean space of dimension `findim ℝ E - 1`. -/
+instance : charted_space (euclidean_space ℝ (fin (findim ℝ E - 1))) (sphere (0:E) 1) :=
+{ atlas            := {f | ∃ v : (sphere (0:E) 1), f = stereographic' v},
+  chart_at         := λ v, stereographic' (-v),
+  mem_chart_source := λ v, by simpa using ne_neg_of_mem_unit_sphere ℝ v,
+  chart_mem_atlas  := λ v, ⟨-v, rfl⟩ }
```

Comment:
A frustrating point about this definition is that everything gets tied up with the literal algebraic expression `findim ℝ E - 1`.  For example, the first of the following examples works, the second fails.
```
example : charted_space (euclidean_space ℝ (fin (findim ℝ ℂ - 1))) (sphere (0:ℂ) 1) :=
metric.sphere.charted_space

example : charted_space (euclidean_space ℝ (fin 1)) (sphere (0:ℂ) 1) :=
metric.sphere.charted_space
```

Reply:
One solution I see is to change this instance to a `def`:
```
def metric.sphere.charted_space_aux
  {E : Type*} [inner_product_space ℝ E] [finite_dimensional ℝ E]
  {E' : Type*} [inner_product_space ℝ E'] [finite_dimensional ℝ E']
  (h : findim ℝ E' = findim ℝ E - 1) :
  charted_space E' ↥(sphere 0 1) :=
```
and then have instances for the important cases only:
```
instance : charted_space (euclidean_space ℝ (fin 1)) (sphere (0:ℂ) 1) :=
metric.sphere.charted_space_aux (by simp)

instance (n : ℕ) : 
  charted_space (euclidean_space ℝ (fin n)) (sphere (0:(euclidean_space ℝ (fin (n + 1)))) 1) :=
metric.sphere.charted_space_aux (by simp)
```

Reply:
Come to think of it, it seems like it would do no harm to have all three instances:
- `charted_space (euclidean_space ℝ (fin (findim ℝ E - 1))) (sphere (0:E) 1)`
- `charted_space (euclidean_space ℝ (fin 1)) (sphere (0:ℂ) 1)`
- `charted_space (euclidean_space ℝ (fin n)) (sphere (0:(euclidean_space ℝ (fin (n + 1)))) 1)`
 
Since the second two are for `fin` of an expression not defeq to `findim ℝ E - 1`, I presume Lean will not see them as duplicate instances to the first one.

However, there remains the question of how to construct all these instances without also duplicating all code and lemmas ...?

Reply:
I guess an "edge case" question with the code that's there now is that in the "bad subtraction" case 0 - 1 = 0, do you definitely want the sphere in R^0 to be a manifold locally modelled on R^0? 

But in general I think that you can't make enough instances to deal with all your use cases. For example I think that even with all the instances above,

```
example (n : ℕ) : charted_space (euclidean_space ℝ (fin n))
  (sphere (0 : euclidean_space ℝ (fin (1 + n))) 1) := by apply_instance
```
would fail. The problem is that `fin (1 + n)` and `fin (n + 1)` are equal but not defeq.

How about
```
def (n : ℕ) (hn : n = findim ℝ E - 1) :
  charted_space (euclidean_space ℝ (fin n)) (sphere (0:E) 1) := ...
```
? Or `hn : n + 1 = findim ℝ E` if you want the edge case to be treated differently. But then you can't make it an instance because type class inference will never find `hn`. In other parts of the library this is treated with the `fact` trick:

```
instance (n : ℕ) [fact (n = findim ℝ E - 1)] :
  charted_space (euclidean_space ℝ (fin n)) (sphere (0:E) 1) :=
```
and then you have to insert the proof yourself into the inference system but at least the instance triggers. I think.

Reply:
In the latest revision I've implemented this `fact` idea of @kbuzzard and it works quite nicely.  With the code I'm about to push, you can get the important "explicit" sphere instances in a few lines:
```lean
def foo : fact (findim ℝ ℂ = 1 + 1) := by simp
local attribute [instance] foo

instance : charted_space (euclidean_space ℝ (fin 1)) (sphere (0:ℂ) 1) := by apply_instance

def foo' (n : ℕ) : fact (findim ℝ (euclidean_space ℝ (fin (n + 1))) = n + 1) := by simp
local attribute [instance] foo'

instance (n : ℕ) :
  charted_space (euclidean_space ℝ (fin n)) (sphere (0 : euclidean_space ℝ (fin (n + 1))) 1) :=
by apply_instance
```
I'd like advice on whether to include these constructions in this file -- or whether instead they should be declared only by the end user in the place where they're used.  And if they're included in this file, where should the `foo`-style facts go?  For example, does 
```lean
def foo : fact (findim ℝ ℂ = 1 + 1) := by simp
```
belong here, or perhaps in `data.complex.module` right after `findim_real_complex`?